fvm/src/kernel/default.rs

```diff
@@ -864,6 +874,95 @@ where
+        if self.read_only {
+            return Err(
+                syscall_error!(ReadOnly, "upgrade_actor cannot be called while read-only").into(),
+            );
+        }
+
+        // Load parameters.
+        let params = if params_id == NO_DATA_BLOCK_ID {
+            None
+        } else {
+            Some(self.blocks.get(params_id)?.clone())
+        };
+
+        let result = self.call_manager.with_transaction(|cm| {
+            let origin = cm.origin();
+
+            let state = cm
+                .get_actor(self.actor_id)?
+                .ok_or_else(|| syscall_error!(NotFound; "actor not found"))?;
+
+            // store the code cid of the calling actor before running the upgrade entrypoint
+            // in case it was changed (which could happen if the target upgrade entrypoint
+            // sent a message to this actor which in turn called upgrade)
+            let code = state.code;
+
+            // update the code cid of the actor to new_code_cid
+            cm.set_actor(
+                origin,
+                ActorState::new(
+                    new_code_cid,
+                    state.state,
+                    state.balance,
+                    state.sequence,
+                    None,
+                ),
+            )?;
+
+            // run the upgrade entrypoint
+            let result = cm.call_actor::<Self>(
+                self.caller,
+                Address::new_id(self.actor_id),
+                Entrypoint::Upgrade(UpgradeInfo { old_code_cid: code }),
+                params,
+                &TokenAmount::from_whole(0),
+                None,
+                false,
+            )?;
+
+            if result.exit_code == ExitCode::OK {
+                // after running the upgrade, our code cid must not have changed
+                let code_after_upgrade = cm
+                    .get_actor(self.actor_id)?
+                    .ok_or_else(|| syscall_error!(NotFound; "actor not found"))?
+                    .code;
+                if code != code_after_upgrade {
+                    return Err(syscall_error!(Forbidden; "re-entrant upgrade detected").into());
+                }
+            }
+
+            Ok(result)
+        });
+
+        match result {
+            Ok(InvocationResult {
+                exit_code: ExitCode::OK,
+                value,
+            }) => {
+                let block_id = match value {
+                    None => NO_DATA_BLOCK_ID,
+                    Some(block) => self.blocks.put_reachable(block).unwrap_or_else(|_| {
+                        log::error!("failed to write to kernel block registry");
+                        NO_DATA_BLOCK_ID
+                    }),
+                };
+                Err(ExecutionError::Abort(Abort::Exit(
+                    ExitCode::OK,
+                    "actor upgraded".to_string(),
+                    block_id,
+                )))
```

Comment:
Don't return an error here, return success. _Then_, in the syscall binding, we convert to an "abort" to unwind through the calling actor.

Otherwise, we need to infect everything with a new error type that's rarely relevant.

Reply:
I am not totally sure I follow, so I can change this to `Ok(ExitCode::OK)` but then how do I include the block_id as that should be returned to the caller on success? Also, are you suggesting that we handle the conversion in the fvm/src or sdk/src?, also what exactly do you mean by "unwind through the calling actor"?

Reply:
Hm.... I'll play with this a bit.

Reply:
> also what exactly do you mean by "unwind through the calling actor"?

Long jump. I.e., instead of returning to the calling actor, we unwind the calling actor's stack.

---

Motivation:

1. My main motivation is that `Abort` ends up everywhere, and it only makes sense in this one case.
2. The syscall succeeded, so returning an error here is kind of strange. We do use errors to unwind/abort actors, but I prefer to keep the kernel a bit isolated from details like that where possible.

Reply:
Ok, so, my idea of using `Abort` was a dud because abort always means "abort this actor on error" and we _don't_ want to abort on syscall errors. I.e., we need to be able to return syscall errors to the actor.

We _could_ add an `Exit` variant to `ExecutionError`. Kind of like your `Abort` solution, but it doesn't have duplicate the `OutOfGas` and/or `Fatal` conditions.

However, this case isn't an error so I'd prefer not to. We're using `Abort` as an from syscall bindings, but that's kind of a special case.

Reply:
Hm. No, `::Exit` is even worse...

Reply:
https://github.com/filecoin-project/ref-fvm/pull/1910